### ISSUE TITLE
GenericObjective.compute() arguments error

### ISSUE THREAD

**Issue by landreman** [PlasmaControl/DESC#388]
When `GenericObjective` is used in an optimization, an error results, `TypeError: GenericObjective.compute() takes 1 positional argument but 3 were given`, where the number 3 might be different depending on the quantity `f`. The same error is produced by this minimal example:
~~~~py
from desc.equilibrium import Equilibrium
from desc.objectives import ObjectiveFunction, GenericObjective, AspectRatio

eq = Equilibrium()

obj = ObjectiveFunction(AspectRatio(), eq=eq)
print("Aspect ratio objective:", obj.compute_scalar(obj.x(eq)))
# So far, everything works fine. Now try to evaluate a GenericObjective the same way:

obj = ObjectiveFunction(GenericObjective("a"), eq=eq)
# The next line gives an error:                                                                                                         
print("Minor radius objective:", obj.compute_scalar(obj.x(eq)))
~~~~
Output:
~~~~
DESC version 0.6.4+129.g0dd97f43,using JAX backend, jax version=0.3.15, jaxlib version=0.3.15, dtype=float64
Using device: CPU, with 5.12 GB available memory
Building objective: aspect ratio
Precomputing transforms
Aspect ratio objective: 31.999999999999986
Building objective: generic
Traceback (most recent call last):
  File "/Users/mattland/Library/CloudStorage/Box-Box/work23/./20230102-01-desc_generic_objective_bug_minimal_example", line 14, in <module>
    print("Minor radius objective:", obj.compute_scalar(obj.x(eq)))
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/_src/api.py", line 526, in cache_miss
    out_flat = xla.xla_call(
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/core.py", line 1919, in bind
    return call_bind(self, fun, *args, **params)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/core.py", line 1935, in call_bind
    outs = top_trace.process_call(primitive, fun_, tracers, params)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/core.py", line 687, in process_call
    return primitive.impl(f, *tracers, **params)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/_src/dispatch.py", line 199, in _xla_call_impl
    compiled_fun = xla_callable(fun, device, backend, name, donated_invars,
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/linear_util.py", line 295, in memoized_fun
    ans = call(fun, *args)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/_src/dispatch.py", line 248, in _xla_callable_uncached
    return lower_xla_callable(fun, device, backend, name, donated_invars, False,
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/_src/profiler.py", line 294, in wrapper
    return func(*args, **kwargs)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/_src/dispatch.py", line 293, in lower_xla_callable
    jaxpr, out_type, consts = pe.trace_to_jaxpr_final2(
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/_src/profiler.py", line 294, in wrapper
    return func(*args, **kwargs)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/interpreters/partial_eval.py", line 2167, in trace_to_jaxpr_final2
    jaxpr, out_type, consts = trace_to_subjaxpr_dynamic2(fun, main)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/interpreters/partial_eval.py", line 2117, in trace_to_subjaxpr_dynamic2
    ans = fun.call_wrapped(*in_tracers_)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/linear_util.py", line 168, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/Users/mattland/DESC/DESC/desc/objectives/objective_funs.py", line 227, in compute_scalar
    f = jnp.sum(self.compute(x) ** 2) / 2
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/_src/api.py", line 526, in cache_miss
    out_flat = xla.xla_call(
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/core.py", line 1919, in bind
    return call_bind(self, fun, *args, **params)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/core.py", line 1935, in call_bind
    outs = top_trace.process_call(primitive, fun_, tracers, params)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/interpreters/partial_eval.py", line 1824, in process_call
    jaxpr, out_type, consts = trace_to_subjaxpr_dynamic2(f, self.main)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/interpreters/partial_eval.py", line 2117, in trace_to_subjaxpr_dynamic2
    ans = fun.call_wrapped(*in_tracers_)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/linear_util.py", line 168, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/Users/mattland/DESC/DESC/desc/objectives/objective_funs.py", line 206, in compute
    [
  File "/Users/mattland/DESC/DESC/desc/objectives/objective_funs.py", line 207, in <listcomp>
    obj.compute(*self._kwargs_to_args(kwargs, obj.args))
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/_src/api.py", line 526, in cache_miss
    out_flat = xla.xla_call(
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/core.py", line 1919, in bind
    return call_bind(self, fun, *args, **params)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/core.py", line 1935, in call_bind
    outs = top_trace.process_call(primitive, fun_, tracers, params)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/interpreters/partial_eval.py", line 1824, in process_call
    jaxpr, out_type, consts = trace_to_subjaxpr_dynamic2(f, self.main)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/interpreters/partial_eval.py", line 2117, in trace_to_subjaxpr_dynamic2
    ans = fun.call_wrapped(*in_tracers_)
  File "/Users/mattland/opt/miniconda3/envs/20220806-03/lib/python3.10/site-packages/jax/linear_util.py", line 168, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
jax._src.traceback_util.UnfilteredStackTrace: TypeError: GenericObjective.compute() takes 1 positional argument but 3 were given

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/mattland/Library/CloudStorage/Box-Box/work23/./20230102-01-desc_generic_objective_bug_minimal_example", line 14, in <module>
    print("Minor radius objective:", obj.compute_scalar(obj.x(eq)))
  File "/Users/mattland/DESC/DESC/desc/objectives/objective_funs.py", line 227, in compute_scalar
    f = jnp.sum(self.compute(x) ** 2) / 2
  File "/Users/mattland/DESC/DESC/desc/objectives/objective_funs.py", line 206, in compute
    [
  File "/Users/mattland/DESC/DESC/desc/objectives/objective_funs.py", line 207, in <listcomp>
    obj.compute(*self._kwargs_to_args(kwargs, obj.args))
TypeError: GenericObjective.compute() takes 1 positional argument but 3 were given
~~~~
I couldn't figure out how to fix that.

Once that issue is resolved, we might also want `GenericObjective` to handle `dim=0` quantities separately from `dim=1` quantities, setting `self._dim_f = 1` and not multiplying by `self.grid.weights` in `compute()`.

**Comment by f0uriest** [comment]
We hadn't used `GenericObjective` much so I think this has been broken for a while, due to the way the `ObjectiveFunction` passes arguments down to the individual objectives and not knowing ahead of time what arguments the `GenericObjective` should take.

This should be fixed now in #391, along with modifications to correctly set the `dim_f` and `scalar` attributes and weighting for scalar values.

**Comment by landreman** [comment]
Thanks! I'll use this for targeting a desired minor radius and average |B|, etc.
